### PR TITLE
[GO-76] added gracefull component

### PR DESCRIPTION
**Pull request by jenchik**

### CODE REVIEW

File: examples/web/main.go [hunk 1]
```diff
@@ -1,20 +1,15 @@
...
+	"github.com/cryptopay-dev/yaga/logger/zap"
 	"github.com/cryptopay-dev/yaga/web"
 )
 
 func main() {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	log := nop.New()
+	log := zap.New("dev")
```
Comment (im-kulikov) [inline examples/web/main.go]: zap.Dev? constant

**Review by im-kulikov — CHANGES_REQUESTED** [review]
that's not what we need now.

**Review by pcherednichenko — APPROVED** [review]

File: web/web.go [hunk 2]
```diff
@@ -132,19 +132,17 @@ func New(opts Options) (*Engine, error) {
...
-// StartAsync HTTP with custom address and return stop channel.
-func StartAsync(e *Engine, bind string) <-chan os.Signal {
-	ch := make(chan os.Signal, 1)
-	signal.Notify(ch, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGABRT)
-
-	go func() {
-		if err := Start(e, bind); err != nil {
+// StartAsync HTTP with custom address and return context.
+func StartAsync(e *Engine, bind string) context.Context {
+	g, ctx := gracefull.NewNotify(context.Background(), e.Logger)
```
Comment (im-kulikov) [inline web/web.go]: you're srsly?

**Comment by jenchik** [comment]
example

```
g, ctx := gracefull.NewNotify(context.Background(), a.Logger)
g.Go(func() error {
        <-ctx.Done()
        a.Workers.Cancel()
        return a.Shutdown(ctx)
    })
if err := a.Group.Wait(); err != nil {
//...
}
```

**Comment by jenchik** [comment]
```
g := gracefull.NewNotify(context.Background(), a.Logger)
g.Go(func(ctx context.Context) error {
        <-ctx.Done()
        a.Workers.Cancel()
        return a.Shutdown(ctx)
    })
if err := a.Group.Wait(); err != nil {
//...
}
```

File: gracefull/gracefull.go [hunk 3]
```diff
@@ -21,38 +22,68 @@ type logger interface {
 }
 
 type gracefull struct {
```
Comment (im-kulikov) [inline gracefull/gracefull.go]: `graceful`.. not `gracefull`

File: graceful/graceful.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	g.cancel()
+}
+
+func (g *graceful) Go(job func(context.Context) error) {
+	f := func() (err error) {
+		defer func() {
+			if r := recover(); r != nil {
+				switch x := r.(type) {
+				case string:
+					err = errors.New(x)
```
Comment (im-kulikov) [inline graceful/graceful.go:40]: errors.Wrap mb?

File: graceful/graceful.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+				case error:
+					err = x
+				default:
+					err = errors.New("Unknown panic")
```
Comment (im-kulikov) [inline graceful/graceful.go:44]: errors.Wrap mb?

File: web/web.go [hunk 6]
```diff
@@ -132,19 +131,17 @@ func New(opts Options) (*Engine, error) {
 	return e, nil
 }
 
-// StartAsync HTTP with custom address and return stop channel.
-func StartAsync(e *Engine, bind string) <-chan os.Signal {
-	ch := make(chan os.Signal, 1)
-	signal.Notify(ch, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGABRT)
-
+// StartAsync HTTP with custom address.
+func StartAsync(e *Engine, bind string, cancel context.CancelFunc) {
```
Comment (im-kulikov) [inline web/web.go]: O_o.. we've discussed this.. it was a completely different story..

Comment (im-kulikov) [reply web/web.go]: `StartAsync(*Engine, bind string) return func(Context) error`...

Comment (jenchik) [reply web/web.go]: than this option is bad?

**Comment by pcherednichenko** [comment]
conflicts here :-(

**Review by pcherednichenko — APPROVED** [review]

**Comment by pcherednichenko** [comment]
а по Wrap ошибок что?

**Comment by jenchik** [comment]
wrapped up

File: graceful/graceful.go (hunk 5 above)
Comment (jenchik) [reply graceful/graceful.go:44]: fixed

File: graceful/graceful.go (hunk 4 above)
Comment (jenchik) [reply graceful/graceful.go:40]: fixed

**Review by im-kulikov — APPROVED** [review]
pls, write messages in English

**Pull request merged**